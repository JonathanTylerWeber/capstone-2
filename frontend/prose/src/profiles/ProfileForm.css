.profile-h1 {
  margin-top: 3em;
  font-weight: bold;
  font-size: 3em;
}