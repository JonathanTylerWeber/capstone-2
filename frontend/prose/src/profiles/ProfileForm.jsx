import React, { useState, useEffect } from "react";
import { useParams, Navigate } from "react-router-dom";
import ProseApi from '../API/api';
import "./ProfileForm.css"
import { Container, Form, Button } from "react-bootstrap"

function ProfileForm() {
  const [user, setUser] = useState(null);
  const [updated, setUpdated] = useState(false);
  const [formData, setFormData] = useState({
    email: "",
    password: ""
  });
  const { username } = useParams();

  useEffect(() => {
    const fetchUserData = async () => {
      try {
        const userData = await ProseApi.getCurrentUser(username);
        setUser(userData);
        setFormData({
          email: userData.email,
          password: ""
        });
      } catch (error) {
        console.error("Error fetching user data:", error);
      }
    };
    fetchUserData();
  }, [username]);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((prevData) => ({
      ...prevData,
      [name]: value
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const { email, password } = formData;
      const data = { email };
      if (password.trim() !== "") {
        data.password = password;
      }
      await ProseApi.updateUser(username, data);
      setUpdated(true);
    } catch (error) {
      console.error("Error updating user:", error);
    }
  };

  if (updated) {
    return <Navigate to="/" />;
  }

  if (!user) {
    return <p>Loading...</p>;
  }

  console.log

  return (
    <Container className="cntr">
      <h1 className="profile-h1">Edit {username}'s Profile</h1>
      <Form onSubmit={handleSubmit} >
        <Form.Group className="mb-3" controlId="email">
          <Form.Label className='label'>Update Email:</Form.Label>
          <Form.Control
            type="email"
            name="email"
            value={formData.email}
            onChange={handleChange}
            required />
        </Form.Group>
        <Form.Group className="mb-3" controlId="password">
          <Form.Label className='label'>New Password:</Form.Label>
          <Form.Control
            type="password"
            name="password"
            placeholder="Enter new password"
            value={formData.password}
            onChange={handleChange}
          />
        </Form.Group>
        <Button className="btn auth-btn" variant="primary" type="submit">
          Save Changes
        </Button>
      </Form>
    </Container>
  );
}

export default ProfileForm;
