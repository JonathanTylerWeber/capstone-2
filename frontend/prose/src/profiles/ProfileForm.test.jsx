import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import ProfileForm from './ProfileForm';
import ProseApi from '../API/api';

vi.mock('../API/api');

const mockUserData = {
  username: 'testuser',
  email: 'testuser@example.com',
};

describe('ProfileForm', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it("matches snapshot", function () {
    const { asFragment } = render(<ProfileForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the component and fetches user data', async () => {
    ProseApi.getCurrentUser.mockResolvedValue(mockUserData);

    render(
      <MemoryRouter initialEntries={['/profile/testuser']}>
        <Routes>
          <Route path="/profile/:username" element={<ProfileForm />} />
        </Routes>
      </MemoryRouter>
    );

    expect(screen.getByText(/Loading.../i)).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.getByLabelText(/Update Email:/i)).toHaveValue(mockUserData.email);
    });

    expect(screen.getByLabelText(/New Password:/i)).toBeInTheDocument();
  });

  it('submits the form with updated email and password', async () => {
    ProseApi.getCurrentUser.mockResolvedValue(mockUserData);
    ProseApi.updateUser.mockResolvedValue({});

    render(
      <MemoryRouter initialEntries={['/profile/testuser']}>
        <Routes>
          <Route path="/profile/:username" element={<ProfileForm />} />
        </Routes>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByLabelText(/Update Email:/i)).toHaveValue(mockUserData.email);
    });

    fireEvent.change(screen.getByLabelText(/Update Email:/i), { target: { value: 'newemail@example.com' } });
    fireEvent.change(screen.getByLabelText(/New Password:/i), { target: { value: 'newpassword' } });
    fireEvent.click(screen.getByRole('button', { name: /Save Changes/i }));

    await waitFor(() => {
      expect(ProseApi.updateUser).toHaveBeenCalledWith('testuser', {
        email: 'newemail@example.com',
        password: 'newpassword',
      });
    });

    await waitFor(() => {
      expect(screen.queryByText(/Loading.../i)).not.toBeInTheDocument();
    });
  });

  it('submits the form with updated email only', async () => {
    ProseApi.getCurrentUser.mockResolvedValue(mockUserData);
    ProseApi.updateUser.mockResolvedValue({});

    render(
      <MemoryRouter initialEntries={['/profile/testuser']}>
        <Routes>
          <Route path="/profile/:username" element={<ProfileForm />} />
        </Routes>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByLabelText(/Update Email:/i)).toHaveValue(mockUserData.email);
    });

    fireEvent.change(screen.getByLabelText(/Update Email:/i), { target: { value: 'newemail@example.com' } });
    fireEvent.change(screen.getByLabelText(/New Password:/i), { target: { value: '' } });
    fireEvent.click(screen.getByRole('button', { name: /Save Changes/i }));

    await waitFor(() => {
      expect(ProseApi.updateUser).toHaveBeenCalledWith('testuser', {
        email: 'newemail@example.com',
      });
    });

    await waitFor(() => {
      expect(screen.queryByText(/Loading.../i)).not.toBeInTheDocument();
    });
  });

  it('navigates to the home page after successful update', async () => {
    ProseApi.getCurrentUser.mockResolvedValue(mockUserData);
    ProseApi.updateUser.mockResolvedValue({});

    render(
      <MemoryRouter initialEntries={['/profile/testuser']}>
        <Routes>
          <Route path="/profile/:username" element={<ProfileForm />} />
          <Route path="/" element={<div>Home Page</div>} />
        </Routes>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByLabelText(/Update Email:/i)).toHaveValue(mockUserData.email);
    });

    fireEvent.change(screen.getByLabelText(/Update Email:/i), { target: { value: 'newemail@example.com' } });
    fireEvent.change(screen.getByLabelText(/New Password:/i), { target: { value: 'newpassword' } });
    fireEvent.click(screen.getByRole('button', { name: /Save Changes/i }));

    await waitFor(() => {
      expect(screen.queryByText(/Loading.../i)).not.toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByText(/Home Page/i)).toBeInTheDocument();
    });
  });
});
