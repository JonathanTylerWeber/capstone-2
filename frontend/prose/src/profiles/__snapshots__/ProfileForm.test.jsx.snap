// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileForm > matches snapshot 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
